src/libsync/vfs/cfapi/cfapiwrapper.cpp

```diff
@@ -124,21 +124,75 @@ void CALLBACK cfApiFetchDataCallback(const CF_CALLBACK_INFO *callbackInfo, const
     }
 
     qint64 offset = 0;
+    QByteArray protrudingData;
+    bool protrudingDataLeft = false;
 
     QObject::connect(&socket, &QLocalSocket::readyRead, &loop, [&] {
-        auto data = socket.readAll();
+        const auto data = socket.readAll();
         if (data.isEmpty()) {
             qCWarning(lcCfApiWrapper) << "Unexpected empty data received" << requestId;
             sendTransferError();
             loop.quit();
             return;
         }
-        sendTransferInfo(data, offset);
-        offset += data.length();
+
+        QByteArray dataToSent;
+        // CFAPI expectes sended blocks to be of a specified size.
+        // Only the last sended block is allowed to be smaller than the specified block size
+        // and therefore marks the end of the transmission.
+        static constexpr auto CFAPI_BLOCK_SIZE = 4096;
+
+        int dataIndex = 0;
+        // Is there any data left from last run?
+        if (!protrudingData.isEmpty()) {
+            QByteArray dataToSent;
+            dataToSent.append(protrudingData);
+
+            // Fill up remaining first block if possible
+            const auto fillupSize = CFAPI_BLOCK_SIZE - dataToSent.size();
+            if (fillupSize > 0) {
+                dataToSent.append(data.left(fillupSize));
+                dataIndex += fillupSize;
+            }
+
+            // Send the block
+            qDebug(lcCfApiWrapper) << "Send protruding data that is left from last run. Size:" << dataToSent.size();
+            sendTransferInfo(dataToSent, offset);
+            offset += dataToSent.length();
+        }
+
+        // Do we still have data to sent?
+        if (dataIndex < data.size()) {
+            const auto blocksCount = (data.size() - (dataIndex + 1)) / CFAPI_BLOCK_SIZE;
+            if (blocksCount != 0) {
+                // Send the blocks one by one
+                for (int i = 0; i < blocksCount; ++i) {
+                    auto dataToSent = data.mid(dataIndex, CFAPI_BLOCK_SIZE);
+                    qDebug(lcCfApiWrapper) << "Send data block. Size:" << dataToSent.size();
+                    sendTransferInfo(dataToSent, offset);
+                    offset += dataToSent.length();
+                    dataIndex += CFAPI_BLOCK_SIZE;
+                }
+            }
+
+            // Do we still have data left? If yes, save this data for the next run and send it then
+            protrudingDataLeft = false;
+            if (dataIndex < data.size()) {
+                protrudingDataLeft = true;
+                protrudingData = data.mid(dataIndex, data.size() - dataIndex);
```

Comment:
@FlexW Won't `data.right(data.size() - dataIndex)` achieve the same? It is more readable and less chance for human error IMHO.

Reply:
Yes, you're right:)